### PR TITLE
DRAFT: RFC for Resource#current_value

### PR DESCRIPTION
**Pull request by jkeiser**
CC @coderanger @danielsdeleo @someara @lamont-granquist @jaym

Until we've worked through the use cases @coderanger has brought up (I haven't had time this week), this is on hold.

### CODE REVIEW

File: new/rfc-template.md [hunk 1]
```diff
@@ -1,11 +1,10 @@
 ---
 RFC: unassigned
-Author: Alan Smithee <asmithee@example.com>
+Author: John Keiser <jkeiser@chef.io
```
Comment (coderanger) [inline new/rfc-template.md]: Probably not intended to commit this.

**Comment by coderanger** [comment]
I don't know if this should be a global API. I agree the general idea is useful, but I'm not sure this needs to be in core. The examples with core resources feel kind of contrived, this really seems to come up more in super-advanced use cases like chef-provisioning and [poise-ruby/python](https://github.com/poise/poise-ruby/blob/master/lib/poise_ruby/resources/ruby_runtime.rb#L58) (if we extend this to the general case of "resources having output data". For the example using `gem_package` to work correctly it would be relying heavily on cloning since you aren't setting `gem_binary` and that might have been used earlier in a previous definition of the resource. This would basically mean this cannot be used safely in library cookbooks as a resource needs to be defined twice, once for the collection and once for output data.

Basically this feels like a footgun unless we drop down a level and come up with a more general solution for resources having output data, of which current state will be a specific use case.

**Comment by coderanger** [comment]
Copying from IRC for the future:

1. A way to get a reference version of a resource that does not cause it to also converge
2. A way to get the output of multiple actions that occur on a resource
3. A way to handle the degenerate case of #2 with a single action in a simple way (90% case)

**Comment by coderanger** [comment]
Also a list of use cases we should map out:

1. Current value of `version` for a `gem_package` where the package resource might or might not have a `gem_binary` (control property) defined earlier in the recipe.
2. Getting the AWS ID for a newly created `machine` resource.
3. Getting `stdout` for an `execute` resource.
4. Getting the `ruby_binary` path for a `ruby_runtime` resource.

**Comment by lamont-granquist** [comment]
I'd like to be able to say:

```ruby
package package_installable?("git-core") ? "git-core" : "git"
```

`package_installable?` would then be some sugar around using the package resource to see if the arg was an available package or not.  I want to be able to re-use the package resource so that I don't have to think about yum and apt and everything else myself.  I also don't want to trigger the installation and then rescue or `epic_fail true`.  And this is cleaner than platform/platform_family/platform_version case statements and is really the behavior that I want rather than having to think about which debian/ubuntu/linuxmint platform_versions renamed 'git-core' to 'git'.

File: new/resource-current-value.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
+---
+RFC: unassigned
+Author: John Keiser <jkeiser@chef.io
+Status: Draft
+Type: Standards Track
+---
+
+# Allow Current Value To Be Read From Resource
```
Comment (martinb3) [inline new/resource-current-value.md]: Did you decide to call this `current_resource` instead of `current_value` after writing this title and the paragraph of text under it? I think these might just need to be updated to:

```
# Allow Current Resource to be Read from Resource Collection

Resources get a `current_resource` method, which discovers an existing resource in the resource collection
```

File: new/resource-current-value.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    As a Chef resource developer,
+    I want to be able to expose all the read code I wrote to my users,
+    So I don't have to tell them to copy and paste it.
+
+## Specification
+
+All Resources get a method, `current_resource`, that returns the current value
+as a new resource. It does not cache this value. It returns `nil` when the
+resource does not exist. By default, it is implemented by creating the provider
+for one of the resource's actions, calling `load_current_resource` on that.
```
Comment (martinb3) [inline new/resource-current-value.md]: I was expecting this would be implemented by returning a reference to the existing resource, not to load a new version of it using the provider. I feel like some providers may have (albeit a bad practice) existing state that `load_current_resource` may not capture. Is there a reason to prefer creating a new resource object over searching the resource collection by name and returning the existing resource object?

**Comment by martinb3** [comment]
Just reading the other comments here, and catching up.

I think my use case is a little less elaborate than @coderanger's described one. I'd simply like to see what parameters were set on a resource, by looking it up. For example, if a resource installs some files at a particular path, I'd love to fetch that path later in a recipe in order to some additional work on it.

Regarding @lamont-granquist's example, I couldn't tell if `package_installable?` happens at converge or compile time. That example feels to me like yet another case I hadn't considered -- drawing on a resource to query some information. Perhaps that would be better suited to something like chef-sugar? Querying state doesn't feel idempotent when used in that example.

Overall, I'd be happy with this RFC being implemented, FWIW. :+1: for writing it @jkeiser!

File: new/resource-current-value.md (hunk 2 above)
Comment (jkeiser) [inline new/resource-current-value.md]: The intent is to read the actual value from the system, not from the resource collection, actually.

File: new/resource-current-value.md (hunk 3 above)
Comment (jkeiser) [inline new/resource-current-value.md]: Because the existing resource in the collection is a statement of what the user wants to be true (a desired value), not what is actually true (current_resource).

Comment (martinb3) [inline new/resource-current-value.md]: Ah. I guess I'm sometimes looking for what the desired value is, like in my example below about the path that some files are going to be copied to.

File: new/resource-current-value.md (hunk 2 above)
Comment (martinb3) [inline new/resource-current-value.md]: But you left in `current_value` when the examples further down call the new DSL `current_resource`, right? Or is this proposing both?

File: new/resource-current-value.md (hunk 3 above)
Comment (jkeiser) [inline new/resource-current-value.md]: Yep. There are reasons to want to find out desired state :) This isn't that, but once we've worked through all the issues I suspect we'll have both.

File: new/resource-current-value.md (hunk 2 above)
Comment (jkeiser) [inline new/resource-current-value.md]: Just current_value--which corresponds to the new `load_current_value` API. current_resource might be the right word, but in that case it would still correspond to `Provider.load_current_resource`, an existing API which lets you load the current value of the resource from the system.

File: new/resource-current-value.md (hunk 3 above)
Comment (martinb3) [inline new/resource-current-value.md]: :+1:

**Comment by lamont-granquist** [comment]
So `package_installable?` clearly would need to be compile-time and the method itself would probably be in chef-sugar, but right now there's no good way to implement that in chef-sugar.   This RFC should allow that method to actually get written.   And again the problem is that in order to write that you'd want an abstraction over yum/apt/etc packages and then be able to query state without modifying it.  That is what is accomplished by being able to grab a resource and inspect the load_current_resource state.  Otherwise we have to right some query tools which re-implement a fascade over yum/apt/etc and that's exactly what we already have with the package pseudo-resource, so I want to re-use that.

And querying state like that is perfectly idempotent.  But instead of having awful case statements around platform and platform_version (and having to take into account debian, ubuntu and linuxmint and where exactly the pv on all of those uniquely change) you just duck-type the operating system.   In pseudo-code it has this kind of shape:

```
install_git_core if platform.respond_to?(:install_git_core)
```

For any given platform and platform_version the boolean value it returns is deterministic, and overall its idempotent and convergent, but it allows you to query the operating system to find out how it behaves rather than having to manage a case statement with knowledge about how every platform and platform_version behaves.  Because ultimately I really don't care.  If I have a platform which lets me install 'git-core' then I know I'm on old-debian where 'git' is really 'gnu interactive toolset' and I need to use 'git-core', otherwise I just use 'git' and that works for every case I'm aware of.  If we discover a new platform and platform_versions which have a similar lineage to debian then it'll just work without modification (although the better win is that we already have that platform, which is linuxmint, and i think i'm the only person in the Chef community who ever bothers to convert ubuntu versions to linuxmint versions whenever i find them).

**Comment by lamont-granquist** [comment]
Also -dev vs -devel vs none for header files gets a lot easier:

```ruby
package "foo"
package "foo-dev" if package_installable?("foo-dev")
package "foo-devel" if package_installable?("foo-devel")
```

That'll support fedora, amazon, all the rhels, debian, ubuntu, raspbian, linuxmint and arch at least right out of the box, and has a good shot of solaris, gentoo, etc i think.  All without a stupid case statement around the platform.

**Comment by lamont-granquist** [comment]
There's another use case here, which is minitest-chef-handler.  It already implements this kind of behavior using resources and load_current_resource in the provider in order to generate matchers for minitest.  So if you're writing a rule that a file should have mode 0644 or whatever, internally minitest will create a file resource and call load_current_resource and then pull the mode out.

If we turn this into a public API it will create less breakage for minitest-chef-handler and we should be able to integrate similar chef resources based matchers with rspec.

**Comment by martinb3** [comment]
Maybe idempotent is a poor word choice on my part, but something about that example does not feel declarative about the state of the system. In any case, I withdraw the comment since I can't really justify why it strikes me that way.

My use case would be to look at the desired state of a resource that has already been declared, not the current state of the resource, so I'd love to see a way for us to look up both. I would think that if we looked up the resource in the resource collection, it would contain both the result of load_current_resource data as well as the desired state (new_resource.params).

**Comment by martinb3** [comment]
Perhaps I should just use the existing `#resources` DSL functionality for my use case, and leave this to the other use cases. There's just something cleaner about `package('foo').version` over `resources(:package, 'foo').version`.

**Comment by lamont-granquist** [comment]
Its still also declarative, but it is giving you a much better if-then language to use:  "if you can install git-core then it should be installed" or "if you can install zlib-devel, then it should be installed".  Rather than "if you are on ubuntu older than X, or linuxmint older than Y, or debian older than X, then git-core should be installed", which is tedious.

**Comment by lamont-granquist** [comment]
seems this snuck into the code:

https://github.com/chef/chef/blob/3baa2c1abc07a21acf9dd1430d4b97700c2835a1/lib/chef/resource.rb#L1161-L1176

should probably close this RFC, we could open the RFC again if there were issues with the current_value implementation there.

**Comment by thommay** [comment]
Agreed.

**Comment by lamont-granquist** [comment]
We have load_current_value and current_value, and this RFC is inaccurate as to how they've worked now for years.
